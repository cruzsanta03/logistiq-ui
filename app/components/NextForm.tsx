import React from "react"

type Props = {
  form: {
    email: string
    phone: string
  }
  handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void
  handleBlur?: (e: React.FocusEvent<HTMLInputElement>) => void
  onSubmit?: (result: any) => void
  format: (val: string) => string
}

const NextForm: React.FC<Props> = ({ form, handleChange, handleBlur, onSubmit, format }) => {
  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    const response = await fetch("/api/nextform", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(form),
    });
    const result = await response.json();
    onSubmit?.(result);
  };

  return (
    <form onSubmit={handleSubmit} className="p-4 border rounded bg-white shadow">
      <label htmlFor="email" className="block font-semibold mb-1">Email</label>
      <input
        id="email"
        name="email"
        type="email"
        placeholder="Email"
        value={form.email}
        onChange={handleChange}
        onBlur={handleBlur}
        className="mb-4 border px-2 py-1 rounded w-full"
      />

      <label htmlFor="phone" className="block font-semibold mb-1">Phone</label>
      <input
        id="phone"
        name="phone"
        type="tel"
        placeholder="Phone"
        value={form.phone}
        onChange={handleChange}
        onBlur={handleBlur}
        className="border px-2 py-1 rounded w-full"
      />

      <div className="mt-4 text-sm text-gray-600">
        <strong>Formatted Phone:</strong> {format(form.phone)}
      </div>

      <button type="submit" className="mt-4 bg-blue-600 text-white px-4 py-2 rounded">Submit</button>
    </form>
  )
}

export default NextForm
