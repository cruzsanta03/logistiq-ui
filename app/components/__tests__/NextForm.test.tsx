import React from "react"
import { render, screen } from "@testing-library/react"
import "@testing-library/jest-dom"
import userEvent from "@testing-library/user-event"
import renderer from "react-test-renderer"
import NextForm from "../NextForm"

describe("NextForm", () => {
  const baseProps = {
    form: {
      email: "",
      phone: ""
    },
    handleChange: jest.fn(),
    format: (val: string) => `formatted-${val}`
  }

  it("matches the snapshot", () => {
    const tree = renderer.create(<NextForm {...baseProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("renders all input fields", () => {
    render(<NextForm {...baseProps} />)
    expect(screen.getByPlaceholderText(/email/i)).toBeInTheDocument()
    expect(screen.getByPlaceholderText(/phone/i)).toBeInTheDocument()
  })

  it("calls handleChange on input type", async () => {
    render(<NextForm {...baseProps} />)
    const input = screen.getByPlaceholderText(/email/i)
    await userEvent.type(input, "test@example.com", { delay: 0 })
    expect(baseProps.handleChange).toHaveBeenCalled()
  })

  it("shows validation error on blur for email and phone", async () => {
    const ComponentWithBlur = () => {
      const [form, setForm] = React.useState({ ...baseProps.form })
      const [errors, setErrors] = React.useState<Record<string, string>>({})

      const handleChange = (e: React.ChangeEvent<HTMLInputElement>) =>
        setForm({ ...form, [e.target.name]: e.target.value })

      const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
        if (!e.target.value.trim()) {
          setErrors((prev) => ({ ...prev, [e.target.name]: "Required" }))
        } else {
          const { [e.target.name]: _, ...rest } = errors
          setErrors(rest)
        }
      }

      return (
        <div>
          <NextForm
            form={form}
            handleChange={handleChange}
            handleBlur={handleBlur}
            format={(val) => val}
          />
          {errors.email && <p>{errors.email}</p>}
          {errors.phone && <p>{errors.phone}</p>}
        </div>
      )
    }

    render(<ComponentWithBlur />)

    const emailInput = screen.getByPlaceholderText(/email/i)
    const phoneInput = screen.getByPlaceholderText(/phone/i)

    await userEvent.clear(emailInput)
    await userEvent.tab()
    await userEvent.clear(phoneInput)
    await userEvent.tab()

    expect(await screen.findAllByText("Required")).toHaveLength(2)
  })
})
