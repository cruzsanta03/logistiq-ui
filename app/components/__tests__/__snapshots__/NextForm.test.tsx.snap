// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextForm matches the snapshot 1`] = `null`;
